File: web/src/AccountService.ts
import ServiceBase from "./ServiceBase";
import {
    AccountDefinition,
    AccountInfoResponse,
    AccountProfile,
    ConfirmationMethod,
    LoginRequest,
    LoginResponse,
    ServiceOptions,
    SignInRequest,
    TokenProtection,
} from "./proto";
import * as proto from "./proto";
import { Base64 } from "js-base64";

import type { Client as BrowserClient } from "nice-grpc-web";

export class AccountService extends ServiceBase {
    client: BrowserClient<typeof AccountDefinition>;

    constructor(options?: ServiceOptions) {
        super(options);

        this.client = this.createClient(AccountDefinition);
    }

    /**
     * protect the given profile
     * @param profile The profile to protect with oberon blinding
     * @param securityCode must be utf-8 encoded `UInt8Array`. `string` will be decoded to utf-8.
     */
    public static async protect(
        profile: string | AccountProfile,
        securityCode: string | Uint8Array
    ): Promise<string> {
        securityCode = AccountService.convertToUtf8(securityCode);
        profile = AccountService.convertToProfile(profile);
        let cloned = AccountProfile.fromPartial(profile);
        const result = await ServiceBase.trinsicProvider.blindOberon(
            cloned,
            securityCode
        );
        cloned.authToken = result;
        cloned.protection = TokenProtection.fromPartial({
            enabled: true,
            method: ConfirmationMethod.Other,
        });
        return AccountService.convertToToken(cloned);
    }

    /**
     * unprotect the given profile
     * @param profile The profile to unprotect with oberon blinding
     * @param securityCode must be utf-8 encoded `UInt8Array`. `string` will be decoded to utf-8.
     */
    public static async unprotect(
        profile: string | AccountProfile,
        securityCode: string | Uint8Array
    ): Promise<string> {
        securityCode = AccountService.convertToUtf8(securityCode);
        profile = AccountService.convertToProfile(profile);
        let cloned = AccountProfile.fromPartial(profile);
        const result = await ServiceBase.trinsicProvider.unblindOberon(
            cloned,
            securityCode
        );
        cloned.authToken = result;
        cloned.protection = TokenProtection.fromPartial({
            enabled: false,
            method: ConfirmationMethod.None,
        });
        return AccountService.convertToToken(cloned);
    }

    public static convertToProfile(
        profile: string | AccountProfile
    ): AccountProfile {
        if (typeof profile == "string") {
            return AccountProfile.decode(Base64.toUint8Array(profile));
        }
        return profile;
    }

    public static convertToToken(profile: string | AccountProfile): string {
        if (typeof profile == "string") {
            return profile;
        }
        return Base64.fromUint8Array(
            AccountProfile.encode(profile as AccountProfile).finish(),
            true
        );
    }

    private static convertToUtf8(
        securityCode: string | Uint8Array
    ): Uint8Array {
        if (typeof securityCode == "string") {
            return new TextEncoder().encode(securityCode);
        } else {
            return securityCode;
        }
    }

    public async signIn(
        request: SignInRequest = SignInRequest.fromPartial({})
    ): Promise<string> {
        let response = await this.client.signIn(request);
        const authToken = AccountService.convertToToken(response.profile!);

        // set the auth token as active for the current service instance
        await this.tokenProvider.saveDefault(authToken);
        return authToken;
    }

    public async login(
        request: LoginRequest = LoginRequest.fromPartial({})
    ): Promise<LoginResponse> {
        return this.client.login(request, {
            metadata: await this.buildMetadata(),
        });
    }

    public async loginConfirm(
        challenge: Uint8Array | undefined,
        authCode: string | Uint8Array
    ): Promise<string> {
        if (typeof challenge === "undefined") {
            throw new TypeError("challenge must be a Uint8Array");
        }

        challenge = AccountService.convertToUtf8(challenge);
        authCode = AccountService.convertToUtf8(authCode);
        let digest = await ServiceBase.trinsicProvider.blake3HashRequest(
            authCode
        );

        let response = await this.client.loginConfirm({
            challenge: challenge,
            confirmationCodeHashed: digest,
        });
        if (response.profile === undefined) {
            return "";
        }

        let authToken = AccountService.convertToToken(response.profile!);
        if (response.profile.protection?.enabled ?? false) {
            authToken = await AccountService.unprotect(authToken, authCode);
        }

        // set the auth token as active for the current service instance
        await this.tokenProvider.saveDefault(authToken);
        return authToken;
    }

    public async loginAnonymous(ecosystemId?: string): Promise<string> {
        const request = LoginRequest.fromPartial({
            email: "",
            invitationCode: "",
            ecosystemId: (ecosystemId ||= ""),
        });
        let response = await this.login(request);
        if (response.profile === undefined) {
            throw new Error("undefined profile returned");
        }
        if (response.profile!.protection!.enabled) {
            throw new Error("protected profile returned from login");
        }
        const authToken = AccountService.convertToToken(response.profile!);
        await this.tokenProvider.saveDefault(authToken);
        return authToken;
    }

    public async getInfo(): Promise<AccountInfoResponse> {
        return await this.info();
    }

// BEGIN Code generated by protoc-gen-trinsic. DO NOT EDIT.
// target: /home/runner/work/sdk/sdk/web/src/AccountService.ts

  /** Get account information */
  public async info(): Promise<proto.AccountInfoResponse> {
    let request = proto.AccountInfoRequest.fromPartial({});
    return this.client.info(request, {
      metadata: await this.buildMetadata(proto.AccountInfoRequest.encode(request).finish())
    });
  }
// END Code generated by protoc-gen-trinsic. DO NOT EDIT.
}

File: java/src/main/java/trinsic/services/AccountService.java
package trinsic.services;

import com.google.common.util.concurrent.Futures;
import com.google.common.util.concurrent.ListenableFuture;
import com.google.protobuf.ByteString;
import com.google.protobuf.InvalidProtocolBufferException;
import java.nio.charset.StandardCharsets;
import java.util.Base64;
import java.util.concurrent.ExecutionException;
import java.util.concurrent.Executors;
import org.jetbrains.annotations.NotNull;
import trinsic.okapi.DidException;
import trinsic.okapi.Hashing;
import trinsic.okapi.Oberon;
import trinsic.okapi.security.v1.Security;
import trinsic.sdk.options.v1.Options;
import trinsic.services.account.v1.*;

public class AccountService extends ServiceBase {
  private final AccountGrpc.AccountFutureStub stub;

  public AccountService() {
    this(null);
  }

  public AccountService(Options.ServiceOptions.Builder options) {
    super(options);
    this.stub = AccountGrpc.newFutureStub(this.getChannel());
  }

  public ListenableFuture<String> signIn() {
    return signIn(SignInRequest.getDefaultInstance());
  }

  public ListenableFuture<String> signIn(@NotNull SignInRequest request) {
    if (request.getEcosystemId().isBlank())
      request = SignInRequest.newBuilder(request).setEcosystemId("default").build();
    var response = this.stub.signIn(request);
    return Futures.transform(
        response,
        input -> {
          var token = Base64.getUrlEncoder().encodeToString(input.getProfile().toByteArray());
          this.getOptionsBuilder().setAuthToken(token);
          return token;
        },
        Executors.newSingleThreadExecutor());
  }

  public static String unprotect(String base64Profile, String securityCode)
      throws InvalidProtocolBufferException, DidException {
    var profile =
        AccountProfile.newBuilder().mergeFrom(Base64.getUrlDecoder().decode(base64Profile)).build();
    var request =
        Security.UnBlindOberonTokenRequest.newBuilder()
            .setToken(profile.getAuthToken())
            .addBlinding(ByteString.copyFromUtf8(securityCode))
            .build();
    var result = Oberon.unBlindToken(request);

    profile =
        AccountProfile.newBuilder(profile)
            .setAuthToken(result.getToken())
            .setProtection(
                TokenProtection.newBuilder()
                    .setMethod(ConfirmationMethod.None)
                    .setEnabled(false)
                    .build())
            .build();
    return Base64.getUrlEncoder().encodeToString(profile.toByteArray());
  }

  public static String protect(String base64Profile, String securityCode)
      throws InvalidProtocolBufferException, DidException {
    var profile =
        AccountProfile.newBuilder().mergeFrom(Base64.getUrlDecoder().decode(base64Profile)).build();
    var request =
        Security.BlindOberonTokenRequest.newBuilder()
            .setToken(profile.getAuthToken())
            .addBlinding(ByteString.copyFromUtf8(securityCode))
            .build();
    var result = Oberon.blindToken(request);

    profile =
        AccountProfile.newBuilder(profile)
            .setAuthToken(result.getToken())
            .setProtection(
                TokenProtection.newBuilder()
                    .setMethod(ConfirmationMethod.Other)
                    .setEnabled(true)
                    .build())
            .build();
    return Base64.getUrlEncoder().encodeToString(profile.toByteArray());
  }

  public ListenableFuture<LoginResponse> login(LoginRequest request)
      throws InvalidProtocolBufferException, DidException {
    return stub.login(request);
  }

  public ListenableFuture<String> loginConfirm(ByteString challenge, String authCode)
      throws InvalidProtocolBufferException, DidException {
    var hashed =
        Hashing.blake3_hash(
                trinsic.okapi.hashing.v1.Hashing.Blake3HashRequest.newBuilder()
                    .setData(ByteString.copyFrom(authCode.getBytes(StandardCharsets.UTF_8)))
                    .build())
            .getDigest();

    var request =
        LoginConfirmRequest.newBuilder()
            .setChallenge(challenge)
            .setConfirmationCodeHashed(hashed)
            .build();

    var response = stub.loginConfirm(request);

    return Futures.transform(
        response,
        input -> {
          if (!input.hasProfile()) return null;

          var token = Base64.getUrlEncoder().encodeToString(input.getProfile().toByteArray());

          if (input.getProfile().getProtection().getEnabled()) {
            try {
              token = unprotect(token, authCode);
            } catch (InvalidProtocolBufferException | DidException e) {
              throw new RuntimeException(e);
            }
          }
          this.getOptionsBuilder().setAuthToken(token);
          return token;
        },
        Executors.newSingleThreadExecutor());
  }

  public ListenableFuture<String> loginAnonymous()
      throws InvalidProtocolBufferException, DidException, ExecutionException,
          InterruptedException {
    return loginAnonymous("default");
  }

  public ListenableFuture<String> loginAnonymous(String ecosystemId)
      throws InvalidProtocolBufferException, DidException, ExecutionException,
          InterruptedException {
    var response = login(LoginRequest.newBuilder().setEcosystemId(ecosystemId).build());

    return Futures.transform(
        response,
        input -> {
          if (!input.hasProfile() || input.getProfile().getProtection().getEnabled()) return null;

          var token = Base64.getUrlEncoder().encodeToString(input.getProfile().toByteArray());
          this.getOptionsBuilder().setAuthToken(token);
          return token;
        },
        Executors.newSingleThreadExecutor());
  }

  public ListenableFuture<AccountInfoResponse> getInfo()
      throws InvalidProtocolBufferException, DidException {
    var request = AccountInfoRequest.newBuilder().build();
    return withMetadata(stub, request).info(request);
  }

  // BEGIN Code generated by protoc-gen-trinsic. DO NOT EDIT.
  // target: C:\work\sdk\java\src\main\java\trinsic\services\AccountService.java

  public ListenableFuture<AccountInfoResponse> info(AccountInfoRequest request)
      throws InvalidProtocolBufferException, DidException {
    // TODO - Handle metadata
    return withMetadata(stub, request).info(request);
  }

  public ListenableFuture<ListDevicesResponse> listDevices(ListDevicesRequest request)
      throws InvalidProtocolBufferException, DidException {
    // TODO - Handle metadata
    return withMetadata(stub, request).listDevices(request);
  }

  public ListenableFuture<RevokeDeviceResponse> revokeDevice(RevokeDeviceRequest request)
      throws InvalidProtocolBufferException, DidException {
    // TODO - Handle metadata
    return withMetadata(stub, request).revokeDevice(request);
  }

  public ListenableFuture<AuthorizeWebhookResponse> authorizeWebhook(
      AuthorizeWebhookRequest request) throws InvalidProtocolBufferException, DidException {
    // TODO - Handle metadata
    return withMetadata(stub, request).authorizeWebhook(request);
  }
  // END Code generated by protoc-gen-trinsic. DO NOT EDIT.
}
